File: leetcode/0424-longest-repeating-character-replacement/solution.py
from collections import Counter


class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        counts = Counter()
        left = 0
        max_freq = 0
        best = 0

        for right in range(len(s)):
            counts[s[right]] += 1
            max_freq = max(max_freq, counts[s[right]])

            window_len = right - left + 1
            if window_len - max_freq > k:
                counts[s[left]] -= 1
                left += 1

            best = max(best, right - left + 1)

        return best

File: leetcode/0424-longest-repeating-character-replacement/solution.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int characterReplacement(string s, int k) {
        vector<int> counts(26, 0);
        int left = 0;
        int maxFreq = 0;
        int best = 0;

        for (int right = 0; right < (int)s.size(); right++) {
            int c = s[right] - 'A';
            counts[c]++;
            maxFreq = max(maxFreq, counts[c]);

            int windowLen = right - left + 1;
            if (windowLen - maxFreq > k) {
                counts[s[left] - 'A']--;
                left++;
            }

            best = max(best, right - left + 1);
        }

        return best;
    }
};
